src/nexmark/queries/q18.rs

```diff
@@ -0,0 +1,401 @@
+use super::NexmarkStream;
+use crate::{
+    nexmark::model::{Bid, Event},
+    operator::{FilterMap, Max},
+    Circuit, OrdZSet, Stream,
+};
+
+//
+// Query 18: Find last bid (Not in original suite)
+//
+// What's a's last bid for bidder to auction?
+// Illustrates a Deduplicate query.
+//
+/// ```sql
+/// CREATE TABLE discard_sink (
+///     auction  BIGINT,
+///     bidder  BIGINT,
+///     price  BIGINT,
+///     channel  VARCHAR,
+///     url  VARCHAR,
+///     dateTime  TIMESTAMP(3),
+///     extra  VARCHAR
+/// ) WITH (
+///   'connector' = 'blackhole'
+/// );
+///
+/// INSERT INTO discard_sink
+/// SELECT auction, bidder, price, channel, url, dateTime, extra
+///  FROM (SELECT *, ROW_NUMBER() OVER (PARTITION BY bidder, auction ORDER BY dateTime DESC) AS rank_number
+///        FROM bid)
+///  WHERE rank_number <= 1;
+/// ```
+
+type Q18Stream = Stream<Circuit<()>, OrdZSet<Bid, isize>>;
+
+pub fn q18(input: NexmarkStream) -> Q18Stream {
+    let bids_by_auction_bidder = input.flat_map_index(|event| match event {
+        Event::Bid(b) => Some(((b.auction, b.bidder), b.date_time)),
+        _ => None,
+    });
+
+    let latest_bid_per_auction_bidder = bids_by_auction_bidder
+        .aggregate::<(), _>(Max)
```

Comment:
I think the `join` below can be avoided by aggregating over bids and returning the bid with the largest `date_time` value.  Ideally this requires an `ArgMax` aggregator (returns the record that maximizes the value of a user-defined closure) but you can also do it with `Fold`.

Reply:
True, done. Performance is slightly better for 10M:

```
cargo bench --bench nexmark --features with-nexmark -- --first-event-rate=10000000 --max-events=10000000 --cpu-cores 8 --query q18 --num-event-generators 6 --source-buffer-size 10000 --input-batch-size 40000
   Compiling dbsp v0.1.0 (/home/michael/dev/vmware/database-stream-processor)
    Finished bench [optimized + debuginfo] target(s) in 3m 35s
     Running benches/nexmark/main.rs (target/release/deps/nexmark-733304a45ec676db)
Starting q18 bench of 10000000 events...
10,000,000 / 10,000,000 [======================================================================================================================================================================================] 100 % 1910681.4912/s 0s
┌───────┬────────────┬───────┬─────────┬─────────────────┬──────────────────┬───────────────┬───────────────┬─────────────┐
│ Query │ #Events    │ Cores │ Elapsed │ Cores * Elapsed │ Throughput/Cores │ Total Usr CPU │ Total Sys CPU │ Max RSS(Kb) │
├───────┼────────────┼───────┼─────────┼─────────────────┼──────────────────┼───────────────┼───────────────┼─────────────┤
│ q18   │ 10,000,000 │ 8     │ 5.235s  │ 41.876s         │ 238.798 K/s      │ 32.143s       │ 1.394s        │ 1,999,048   │
└───────┴────────────┴───────┴─────────┴─────────────────┴──────────────────┴───────────────┴───────────────┴─────────────┘
```

But still unable to run with 100M (with the same args)
```
$ cargo bench --bench nexmark --features with-nexmark -- --first-event-rate=10000000 --max-events=100000000 --cpu-cores 8 --query q18 --num-event-generators 6 --source-buffer-size 10000 --input-batch-size 40000 
    Finished bench [optimized + debuginfo] target(s) in 0.10s
     Running benches/nexmark/main.rs (target/release/deps/nexmark-733304a45ec676db)
Starting q18 bench of 100000000 events...
83,200,000 / 100,000,000 [========================================================================================================================================================>------------------------------] 83 % 166423.4625/s 2merror: bench failed, to rerun pass `--bench nexmark`

Caused by:
  process didn't exit successfully: `/home/michael/dev/vmware/database-stream-processor/target/release/deps/nexmark-733304a45ec676db --first-event-rate=10000000 --max-events=100000000 --cpu-cores 8 --query q18 --num-event-generators 6 --source-buffer-size 10000 --input-batch-size 40000 --bench` (signal: 9, SIGKILL: kill)
```